storage-provider/server/src/p2p.rs

```diff
@@ -0,0 +1,238 @@
+use clap::ValueEnum;
+use ed25519_dalek::{pkcs8::DecodePrivateKey, SigningKey};
+use libp2p::{
+    futures::StreamExt, identify, identity::Keypair, rendezvous, rendezvous::Namespace,
+    swarm::SwarmEvent, Multiaddr, PeerId, Swarm,
+};
+use register::{RegisterBehaviour, RegisterBehaviourEvent};
+use serde::de;
+use tokio_util::{sync::CancellationToken, task::TaskTracker};
+use tracing::{error, info};
+
+mod bootstrap;
+mod register;
+
+pub(crate) use bootstrap::BootstrapConfig;
+pub(crate) use register::RegisterConfig;
+
+const P2P_NAMESPACE: &str = "polka-storage";
+
+#[derive(Debug, Clone, Copy, ValueEnum)]
+pub enum NodeType {
+    Bootstrap,
+    Register,
+}
+
+impl Display for NodeType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum P2PError {
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+    #[error(transparent)]
+    DialError(#[from] libp2p::swarm::DialError),
+    #[error("Invalid TCP config for swarm")]
+    InvalidTcpConfig,
+    #[error("Invalid behaviour config for swarm")]
+    InvalidBehaviourConfig,
+    #[error(transparent)]
+    TOMLError(#[from] toml::de::Error),
+    #[error("Failed to register at rendezvous point {0}")]
+    RegistrationFailed(PeerId),
+    #[error(transparent)]
+    P2PTransportError(#[from] libp2p::TransportError<std::io::Error>),
+}
+
+fn deser_keypair<'de, D: de::Deserializer<'de>>(d: D) -> Result<Keypair, D::Error> {
+    let src: String = de::Deserialize::deserialize(d)?;
+    let key = if let Some(stripped) = src.strip_prefix('@') {
+        let path = PathBuf::from_str(stripped)
+            .map_err(de::Error::custom)?
+            .canonicalize()
+            .map_err(de::Error::custom)?;
+        SigningKey::read_pkcs8_pem_file(path).map_err(de::Error::custom)?
+    } else {
+        let hex_key = hex::decode(src).map_err(de::Error::custom)?;
+        SigningKey::try_from(hex_key.as_slice()).map_err(de::Error::custom)?
+    };
+    Keypair::ed25519_from_bytes(key.to_bytes()).map_err(de::Error::custom)
+}
+
+pub async fn run_bootstrap_node(
+    config: BootstrapConfig,
+    token: CancellationToken,
+) -> Result<(), P2PError> {
+    info!("Starting P2P bootstrap node");
+    let tracker = TaskTracker::new();
+    let (swarm, addr) = config.create_swarm()?;
+
+    tokio::select! {
+        res = bootstrap(swarm, addr) => {
+            if let Err(e) = res {
+                error!("Failed to start P2P node. Reason: {e}");
+                return Err(e);
+            }
+        },
+        _ = token.cancelled() => {
+            tracing::info!("P2P node has been stopped by the cancellation token...");
+        },
+    }
+
+    tracker.close();
+    tracker.wait().await;
+
+    Ok(())
+}
+
+pub async fn run_register_node(
+    config: RegisterConfig,
+    token: CancellationToken,
+) -> Result<(), P2PError> {
+    info!("Starting P2P register node");
```

Comment:
Either use `tracing::<level>` or just `<level>` but make it consistent (personally, I prefer the qualified one)